src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- 'hasEdge' x y (vertex (Left x))  == False
+-- 'hasEdge' x y (vertex (Right y)) == False
+-- @
+vertex :: Either a b -> AdjacencyMap a b
+vertex (Left x)  = leftVertex x
+vertex (Right y) = rightVertex y
+
+-- | Constructs the bipartite graph comprising /a single edge/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (edge x y)  == Map.'Map.singleton' x (Set.'Set.singleton' y)
+-- 'rightAdjacencyMap' (edge x y) == Map.'Map.singleton' y (Set.'Set.singleton' x)
+-- 'hasEdge' x y (edge x y)       == True
+-- 'hasEdge' y x (edge x y)       == (x == y)
+-- @
+edge :: a -> b -> AdjacencyMap a b
+edge x y = BAM (Map.singleton x (Set.singleton y)) (Map.singleton y (Set.singleton x))
+
+-- | /Overlay/ two bipartite graphs. This is a commutative, associative and
+-- idempotent operation with the identity 'empty'.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- 'isEmpty'     (overlay x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- @
+overlay :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+overlay (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionWith Set.union lr1 lr2) (Map.unionWith Set.union rl1 rl2)
+
+-- | /Connect/ two bipartite graphs, not adding the inappropriate edges. This
+-- is an associative operation with the identity 'empty', which distributes
+-- over 'overlay' and obeys the decomposition axion.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory. Note that the
+-- number of edges in the resulting graph is quadratic with respect to the
+-- number of vertices in the arguments: /m = O(m1 + m2 + l1 * r2 + l2 * r1)/.
+--
+-- @
+-- 'isEmpty'     (connect x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (connect x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (connect x y) >= 'vertexCount' x
+-- 'vertexCount' (connect x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (connect x y) >= 'edgeCount' x
+-- 'edgeCount'   (connect x y) >= 'edgeCount' y
+-- 'edgeCount'   (connect x y) >= 'leftVertexCount' x * 'rightVertexCount' y
+-- 'edgeCount'   (connect x y) <= 'leftVertexCount' x * 'rightVertexCount' y + 'rightVertexCount' x * 'leftVertexCount' y + 'edgeCount' x + 'edgeCount' y
+-- @
+connect :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+connect (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionsWith Set.union [ lr1, lr2,
+                                                                      Map.fromSet (const $ Map.keysSet rl2) (Map.keysSet lr1),
+                                                                      Map.fromSet (const $ Map.keysSet rl1) (Map.keysSet lr2)]) $
```

Comment:
Please try not to exceed 80 characters per line, if possible. Long lines are difficult to review.